components/proxy/src/main/java/com/hotels/styx/proxy/plugin/PluginsMetadata.java

```diff
@@ -51,20 +53,12 @@
 
         activePluginsNames.forEach(name ->
                 checkArgument(plugins.containsKey(name), "No such plugin '%s' in %s", name, plugins));
-
-        this.activePlugins = activePluginsNames.stream()
-                .map(this::pluginMetadata)
-                .collect(toList());
-    }
-
-    public List<PluginMetadata> activePlugins() {
-        return activePlugins;
     }
 
-    private PluginMetadata pluginMetadata(String name) {
-        PluginMetadata metadata = plugins.get(name);
-
-        return metadata.attachName(name);
+    public List<Pair<String, SpiExtension>> activePlugins() {
```

Comment:
This is supposed to be a generic container of two elements, i.e.  a tuple of two elements. I like your suggestion for `Named<E>` but remember that the other component is not necessarily a `String`. Consider for example a `Pair<Id, Origin>`, etc.

We need something like this because AFAIK we cannot instantiate a `Map.Entry<K,V>`. We often need to use maps as a stream sources, or we need to perform stream operations on map elements. This necessitates containers of two elements. So instead of creating them in ad-hoc manner it would be better to have a generic one in styx common.

Reply:
I understand the intention of a `Pair` class, but I'm not convinced that the benefits outweigh the drawbacks. 

While it may be true that we would need a class for both `String` and `Id`, it doesn't seem likely that there will be much need for keys that do not conform to those classes.

I can't agree that `Pair` can be more descriptive, even if it is more practical, as any data class could hypothetically be replaced with a tuple, yet this is not typically done. 

The use of this class isn't a deal-breaker for the PR, but something to consider.

Reply:
Well, it could equally well be a `Pair<OriginState, List<Origin>>`, `Pair<ProducerState, Int>`, `Pair<Origin, ConnectionPool.Stats>` just to name a few. What is common with them? Just that they are a pair of attributes that are probably used transiently during some stream transformation.

So it is in this transient context IMHO where `Pair` is perfectly descriptive. It is not meant to be a specific business object and therefore not supposed to be used as one. Similarly you wouldn't normally use a `List<T>` as a business object on its own, but it still has a perfectly descriptive name and serves important purpose as a building block in your application's implementation.